util/topsql/stmtstats/stmtstats.go

```diff
@@ -0,0 +1,229 @@
+	// Count more data here.
+}
+
+// OnExecutionFinished implements StatementObserver.OnExecutionFinished.
+func (s *StatementStats) OnExecutionFinished(sqlDigest, planDigest []byte) {
+	// Count more data here.
+}
+
+// GetOrCreateStatementStatsItem creates the corresponding StatementStatsItem
+// for the specified SQLPlanDigest and timestamp if it does not exist before.
+// GetOrCreateStatementStatsItem is just a helper function, not responsible for
+// concurrency control, so GetOrCreateStatementStatsItem is **not** thread-safe.
+func (s *StatementStats) GetOrCreateStatementStatsItem(sqlDigest, planDigest []byte) *StatementStatsItem {
+	key := SQLPlanDigest{SQLDigest: BinaryDigest(sqlDigest), PlanDigest: BinaryDigest(planDigest)}
+	item, ok := s.data[key]
+	if !ok {
+		s.data[key] = NewStatementStatsItem()
+		item = s.data[key]
+	}
+	return item
+}
+
+// AddExecCount is used to count the number of executions of a certain SQLPlanDigest.
+// AddExecCount is thread-safe.
+func (s *StatementStats) AddExecCount(sqlDigest, planDigest []byte, n uint64) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	item := s.GetOrCreateStatementStatsItem(sqlDigest, planDigest)
+	item.ExecCount += n
+}
+
+// AddKvExecCount is used to count the number of executions of a certain SQLPlanDigest for a certain target.
+// AddKvExecCount is thread-safe.
+func (s *StatementStats) AddKvExecCount(sqlDigest, planDigest []byte, target string, n uint64) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	item := s.GetOrCreateStatementStatsItem(sqlDigest, planDigest)
+	item.KvStatsItem.KvExecCount[target] += n
+}
+
+// Take removes all existing StatementStatsMap data from StatementStats.
+// Take is thread-safe.
+func (s *StatementStats) Take() StatementStatsMap {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	data := s.data
+	s.data = StatementStatsMap{}
+	return data
+}
+
+// SetFinished marks StatementStats as "finished".
+// The background goroutine will periodically detect whether each
+// StatementStats has been finished, if so, it will be cleaned up.
+func (s *StatementStats) SetFinished() {
+	s.finished.Store(true)
+}
+
+// Finished returns whether the StatementStats has been finished.
+func (s *StatementStats) Finished() bool {
+	return s.finished.Load()
+}
+
+// BinaryDigest is converted from parser.Digest.Bytes(), and the purpose
+// is to be used as the key of the map.
+type BinaryDigest string
+
+// SQLPlanDigest is used as the key of StatementStatsMap to
+// distinguish different sql.
+type SQLPlanDigest struct {
+	SQLDigest  BinaryDigest
+	PlanDigest BinaryDigest
+}
+
+// StatementStatsMap is the local data type of StatementStats.
+type StatementStatsMap map[SQLPlanDigest]*StatementStatsItem
+
+// Merge merges other into StatementStatsMap.
+// Values with the same SQLPlanDigest will be merged.
+//
+// After executing Merge, some pointers in other may be referenced
+// by m. So after calling Merge, it is best not to continue to use
+// other unless you understand what you are doing.
```

Comment:
I think this looks ok, we only have one place to use `Merge`:
```go
for stat in stats {
     total.Merge(stat.Take())
}
```
And some code will rely on pointer behavior instead of value copy.

Reply:
How about making this Take logic by default? In this way you can prevent future developers (or even yourself) to make mistakes.

Reply:
What is `Making this Take logic by default`? Do you mean that we no longer provide a public `Merge` method, and only merge internally in the place of `stats.Take()` to avoid future misuse?

Reply:
This PR will be merged soon, but we can continue to discuss here, this point can be optimized in the next PR.